keps/sig-node/compute-device-assignment.md

```diff
@@ -0,0 +1,153 @@
+
+## User Journeys
+
+### Device Monitoring Agents
+
+* As a _Cluster Administrator_, I provide a set of devices from various vendors in my cluster.  Each vendor independently maintains their own agent, so I run monitoring agents only for devices I provide.  Each agent adheres to to the [node monitoring guidelines](https://docs.google.com/document/d/1_CdNWIjPBqVDMvu82aJICQsSCbh2BR-y9a8uXjQm4TI/edit?usp=sharing), so I can use a compatible monitoring pipeline to collect and analyze metrics from a variety of agents, even though they are maintained by different vendors.
+* As a _Device Vendor_, I manufacture devices and I have deep domain expertise in how to run and monitor them. Because I maintain my own Device Plugin implementation, as well as Device Monitoring Agent, I can provide consumers of my devices an easy way to consume and monitor my devices without requiring open-source contributions. The Device Monitoring Agent doesn't have any dependencies on the Device Plugin, so I can decouple monitoring from device lifecycle management. My Device Monitoring Agent works by periodically querying the `/pods` endpoint to discover which devices are being used, and to get the container/pod metadata associated with the metrics:
+
+![image](https://user-images.githubusercontent.com/11936386/42943622-e8c676f2-8b80-11e8-8d8f-96a016c6f3b4.png)
+
+### Network Device Plugins
+
+* As a _Device Vendor_, I manufacture network devices.  Even though the Kubelet exposes different interfaces for network setup (CNI), and device lifecycle management (Device Plugin), I can enable consumption of my network devices without requiring communication between my CNI and Device Plugin implementations.  This allows me to independently develop, test, and release my component for network setup, and my component for managing network devices without any dependencies between them. My Device Plugin exposes the set of network devices on the node, but does _not_ perform network setup when `Allocate` is called. Instead, when `CNI_ADD` is called on my CNI plugin, it queries the kubelet's `/pods` endpoint to discover the network devices it should perform network setup for:
+
+![image](https://user-images.githubusercontent.com/11936386/42944024-2fed118e-8b82-11e8-9e45-fd9cd7d78e8c.png)
+
+### Guaranteed Device Scheduling
+
+* As a _Cluster Admin and Application Developer_, I have a workload with strict requirements about the set of devices it is assigned that cannot be expressed with current kubernetes APIs (e.g. numa topology requirements). If my workload is assigned to a node that has the devices I need, but not in the correct configuration, my workloads performance suffers considerably.  Thus, the node's best-effort attempt to provide a "good" set of devices is not enough to meet the needs of my workload.  I can run a custom scheduler to assign devices to containers during scheduling, and can _guarantee_ that the set of devices the workload is assigned meets my requirements.
```

Comment:
This proposal does not expose NUMA topology, or other HW intricacies any more than they are already exposed today.  If anything, this proposal pushes any cluster-level topology solutions out-of-tree.

Users could, in theory, select a specific device for their pod, just as they could select a specific node.  We should document that this is not recommended.

One assumption implicit in this proposal and in the current device plugin design is that devices are assumed to be stateful, and cannot be "swapped" to a different container after being assigned to the initial container.  Thus, making a permanent binding between device and container does not limit us as long as this assumption holds.  The fact that the binding is permanent is also the reason why cluster-level scheduling is appealing: if devices could be swapped, then we could (eventually) expect the kubelet to "shuffle" devices to ensure pods have a reasonable set of devices assigned at all times.   CPUs and memory banks do not have this property, and thus should not be explicitly bound to a container.

My understanding is that topology awareness internal to the kubelet is something the kubernetes community plans to tackle in the future.  I agree that performance gains from device topology awareness rely in part on the kubelet's ability to assign sensible CPUs (assuming they are using CPU pinning).  Initially, this would likely only serve some of the lower-hanging-fruit use-cases, such as directly linked GPU cards.

Reply:
Why don't you place the Device ID in the pod Status then? A pod with no device ID specified can get attached to different nodes across node reboots and that change is OK if device ID is a "Status" property.

Reply:
I consider the fact that device ID cannot change to be a positive, rather than a downside, as it is a simplifying assumption for agents that consume the information.  It would be more complicated to implement, for example, a GPU monitoring agent if had to assume each container could be re-bound to a different GPU.
I was not aware that pods could be reassigned to a different node if the node it is currently on reboots.  Ill have to talk to @bsalamat about that and figure out what the correct behavior is.
I generally prefer spec over status (although I can add it as an alternative) because [status should be observable](https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#spec-and-status) from the node.

Reply:
A pod once bound to a node cannot move across nodes. But the node a pod is currently attached to can be reset and kubelet can assign new devices the set of pods that are already bound to that node as part of readmission. If the device is part of Spec then it is not possible to alter the assignment after a reboot. Perhaps you see device assignment being static as a feature?
Monitoring systems should ideally adapt to changes in device assignment. I don't see why monitoring agents would influence the design of Spec vs Status.